src/instr.rs

```diff
@@ -42,3 +95,149 @@ pub fn interpret_token(token: parser::Token) -> Instr {
+    }
+}
+fn vi_normal_mode(token: parser::Token) -> Instr {
+    match token {
+        parser::Token::Text(text)   => match text.as_ref() {
+            "h"                     => Instr::MoveCursorLeft,
+            "j"                     => Instr::HistoryNext,
+            "k"                     => Instr::HistoryPrev,
+            "l"                     => Instr::MoveCursorRight,
+            "0"                     => Instr::Digit(0),
+            "$"                     => Instr::MoveCursorEnd,
+
+            "x"                     => Instr::DeleteCharRightOfCursor,
+            "s"                     => Instr::Substitute,
+            "r"                     => Instr::ReplaceMode,
+            "c"                     => Instr::ChangeMode,
+            "C"                     => Instr::ChangeToEnd,
+            "d"                     => Instr::DeleteMode,
+            "D"                     => Instr::DeleteToEnd,
+
+            "e"                     => Instr::MoveEndOfWordRight,
+            "E"                     => Instr::MoveEndOfWordWsRight,
+            "w"                     => Instr::MoveWordRight,
+            "W"                     => Instr::MoveWordWsRight,
+            "b"                     => Instr::MoveWordLeft,
+            "B"                     => Instr::MoveWordWsLeft,
+            "t"                     => Instr::MoveCharMode(CharMoveType::BeforeRight),
+            "T"                     => Instr::MoveCharMode(CharMoveType::BeforeLeft),
+            "f"                     => Instr::MoveCharMode(CharMoveType::Right),
+            "F"                     => Instr::MoveCharMode(CharMoveType::Left),
+
+            "a"                     => Instr::Append,
+            "A"                     => Instr::AppendEnd,
+            "i"                     => Instr::Insert,
+            "I"                     => Instr::InsertStart,
+
+            "1"                     => Instr::Digit(1),
+            "2"                     => Instr::Digit(2),
+            "3"                     => Instr::Digit(3),
+            "4"                     => Instr::Digit(4),
+            "5"                     => Instr::Digit(5),
+            "6"                     => Instr::Digit(6),
+            "7"                     => Instr::Digit(7),
+            "8"                     => Instr::Digit(8),
+            "9"                     => Instr::Digit(9),
+
+            _                       => Instr::Noop,
+        },
+        _                           => vi_common(&token),
+    }
+}
+fn vi_replace_mode(token: parser::Token) -> Instr {
+    match token {
+        parser::Token::Text(text)   => Instr::ReplaceAtCursor(text),
+        _                           => Instr::NormalMode,
+    }
+}
+fn vi_move_char_mode(move_type: CharMoveType, token: parser::Token) -> Instr {
+    match token {
+        parser::Token::Text(ref text) if text.chars().count() == 1 => match move_type {
```

Comment:
Unwrap here should be safe as we check and make sure count() == 1, but I'll update it.